java/org/apache/tomcat/util/scan/StandardJarScanner.java

```diff
@@ -123,6 +128,98 @@ public void setJarScanFilter(JarScanFilter jarScanFilter) {
     }
 
     /**
+     * Tests is the file is a zip file
+     *
+     * @param file
+     * @return
+     * @throws IOException
+     */
+    private static boolean isZipFile(File file) throws IOException {
+        if (file.isDirectory()) {
+            return false;
+        }
+        if (!file.canRead()) {
+            throw new IOException("Cannot read file " + file.getAbsolutePath());
+        }
+        if (file.length() < 4) {
+            return false;
+        }
+        try (DataInputStream in = new DataInputStream(new FileInputStream(file))) {
+            return in.readInt() == 0x504b0304;
+        }
+    }
+
+    /**
+     * Returns the URLS of all jars managed directly or indirectly (given
+     * "Class-Path" manifest relationships) by the class loader
+     *
+     * @param cl
+     * @return
+     */
+    public static URL[] getURLs(URLClassLoader cl) {
+        HashSet<URL> visited = new HashSet<>();
+        URL[] urLs = cl.getURLs();
+        if (urLs != null) {
+            for (URL urL : urLs) {
+                try {
+                    visitJar(visited, urL);
+                } catch (IOException iOException) {
+                    throw new RuntimeException(iOException);
+                }
+            }
+        }
+        URL[] ret = new URL[visited.size()];
+        int i = 0;
+        for (URL url : visited) {
+            ret[i++] = url;
+        }
+        return ret;
+    }
+
+    /**
+     * Performs a DFS traversal over the graph of jars given the "Class-Path"
+     * reference
+     *
+     * @param visited
+     * @param jar
+     * @throws IOException
+     */
+    public static void visitJar(Set<URL> visited, URL jar) throws IOException {
+        File file = new File(jar.getFile());
+        if (!file.exists()) {
+            return;
+        }
+        visited.add(jar);
+
+        if (!isZipFile(file)) {
+            return;
+        }
+
```

Comment:
While it isn't as clean, it is probably going to be more efficient to check the manifest at the same time as the scan rather than opening the JAR once to check the manifest and a second time to scan it.